### PR TITLE
CLI command to create/destroy AWS bastion instance

### PR DESCRIPTION
**Pull request by csrwng**
Adds command to CLI to create a bastion machine with a public IP that
allows ssh access to private machines for a cluster.
The command works for Hypershift clusters as well as regular OCP
clusters.

### CODE REVIEW

File: cmd/bastion/aws/create.go [hunk 1]
```diff
@@ -0,0 +1,549 @@
...
+	}
+	cmd.Flags().StringVar(&opts.Namespace, "namespace", opts.Namespace, "The namespace of the hostedcluster")
+	cmd.Flags().StringVar(&opts.Name, "name", opts.Name, "The name of the hostedcluster")
+	cmd.Flags().StringVar(&opts.InfraID, "infra-id", opts.InfraID, "The infra ID to use for creating the bastion")
+	cmd.Flags().StringVar(&opts.Region, "region", opts.Region, "The region to use for creating the bastion")
+	cmd.Flags().StringVar(&opts.SSHKeyFile, "ssh-key-file", opts.SSHKeyFile, "File with public SSH key to use for bastion instance")
+	cmd.Flags().StringVar(&opts.AWSCredentialsFile, "aws-creds", opts.AWSCredentialsFile, "File with AWS credentials")
+	cmd.Flags().BoolVar(&opts.Wait, "wait", opts.Wait, "Wait for instance to be running")
+
+	cmd.MarkFlagRequired("ssh-key-file")
```
Comment (alvaroaleman) [inline cmd/bastion/aws/create.go:56]: Can we maybe default this to the value of the referenced secret on the hosted cluster if set, similiar to region and infraID?

File: cmd/bastion/aws/create.go [hunk 2]
```diff
@@ -0,0 +1,549 @@
...
+		return "", fmt.Errorf("cannot lookup existing subnet: %w", err)
+	}
+	if len(subnetID) == 0 {
+		return "", fmt.Errorf("no public subnet was found")
+	}
+
+	runCtx, cancel := context.WithTimeout(ctx, 2*time.Minute)
+	defer cancel()
+	result, err := ec2Client.RunInstancesWithContext(runCtx, &ec2.RunInstancesInput{
+		ImageId:      aws.String("resolve:ssm:/aws/service/ami-amazon-linux-latest/amzn2-ami-hvm-x86_64-gp2"),
```
Comment (alvaroaleman) [inline cmd/bastion/aws/create.go:392]: Oh wow, TILd

File: cmd/bastion/aws/create.go [hunk 3]
```diff
@@ -0,0 +1,549 @@
...
+			break
+		}
+	}
+	return subnetID, nil
+}
+
+func existingInstance(ctx context.Context, ec2Client *ec2.EC2, infraID string) (string, error) {
+	instanceCtx, cancel := context.WithTimeout(ctx, 2*time.Minute)
+	defer cancel()
+	instanceFilter := []*ec2.Filter{
```
Comment (alvaroaleman) [inline cmd/bastion/aws/create.go:471]: I think I am missing something here, why does this filter not match on all nodes?

File: cmd/bastion/aws/destroy.go [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+		}
+	}
+	return nil
+}
+
+func (o *destroyBastionOpts) Run(ctx context.Context) error {
+
+	var infraID, region string
+
+	if len(o.Name) > 0 {
```
Comment (alvaroaleman) [inline cmd/bastion/aws/destroy.go:94]: I feel like we should make the whole "Default region and infraID if unset from hostedcluster in api", we have it like five times now?

File: cmd/bastion/aws/create.go [hunk 5]
```diff
@@ -0,0 +1,549 @@
...
+				DeviceIndex:              aws.Int64(0),
+				AssociatePublicIpAddress: aws.Bool(true),
+				SubnetId:                 aws.String(subnetID),
+				Groups:                   []*string{aws.String(sgID)},
+			},
+		},
+		TagSpecifications: []*ec2.TagSpecification{
+			{
+				ResourceType: aws.String("instance"),
+				Tags: []*ec2.Tag{
```
[the remark this answers is not in the document]
Comment (csrwng) [reply cmd/bastion/aws/create.go:408]: we do have a `bastion destroy` command and it is identified as a bastion host by its name
`[infra-id]-bastion`. The thinking being that for any infra-id there would only be one bastion. The command is supposed to be idempotent. If you run `bastion create` again and one exists, it only tells you that one already exists.

File: cmd/bastion/aws/create.go [hunk 6]
```diff
@@ -0,0 +1,549 @@
...
+	}
+	for _, instance := range result.Instances {
+		instanceID := aws.StringValue(instance.InstanceId)
+		log.Info("Created ec2 instance", "id", instanceID, "name", instanceName(infraID))
+		return instanceID, nil
+	}
+	return "", fmt.Errorf("no instances were created")
+}
+
+func existingSubnet(ctx context.Context, ec2Client *ec2.EC2, infraID string) (string, error) {
```
[the remark this answers is not in the document]
Comment (csrwng) [reply cmd/bastion/aws/create.go:432]: Yes, just need its ID to pass it to the runInstances call

File: cmd/bastion/aws/destroy.go [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+		Use:          "aws",
+		Short:        "Destroys AWS bastion instance",
+		SilenceUsage: true,
+	}
+
+	cmd.Flags().StringVar(&opts.Namespace, "namespace", opts.Namespace, "The namespace of the hostedcluster")
+	cmd.Flags().StringVar(&opts.Name, "name", opts.Name, "The name of the hostedcluster")
+	cmd.Flags().StringVar(&opts.InfraID, "infra-id", opts.InfraID, "The infra ID to use for creating the bastion")
+	cmd.Flags().StringVar(&opts.Region, "region", opts.Region, "The region to use for creating the bastion")
+	cmd.Flags().StringVar(&opts.AWSCredentialsFile, "aws-creds", opts.AWSCredentialsFile, "File with AWS credentials")
```
[the remark this answers is not in the document]
Comment (csrwng) [reply cmd/bastion/aws/destroy.go:44]: Like I mentioned above, the idea is to only create a single bastion per infra-ID. If you call `create bastion` again with the same infraID, you'll only be told that one exists. You won't create a new one until you call `destroy bastion`.

File: cmd/bastion/aws/destroy.go (hunk 4 above)
Comment (csrwng) [reply cmd/bastion/aws/destroy.go:94]: The issue is that in other commands like `create cluster` and `destroy cluster` we need more than just those 2 like domain

**Comment by csrwng** [comment]
Thx @alvaroaleman for the review. As far as the e2e goes, is it enough that we will excercise this indirectly when we have a failure of nodes to join? After this command, I am going to add another command that uses this one to create the bastion and download gzipped journals from each of the worker machines.

**Pull request merged**